drivers: uart: fix use of port / config info from const struct

jenmwms opened the pull request:
This driver was assigning to a field that was stored in a const
structure. The issue was (mostly) resolved in af43e14 by moving
the field to a RAM structure and in #25462 by preserving the
original const qualifier. Moving the field was not implemented
for PCIE=n and introduced an inline function for the helper
macros that have otherwise been consistent.

This commit further addresses the issue by moving the port field for
PCIE=y and PCIE=n. It lets the get_port() workaround be refactored
back to DEV_DATA() for consistency in the helper macros. The PCIE
probe is moved to the init() function to avoid probing PCIE every
time UART is being configured.

Tested with qemu_x86 and qemu_x86_64 tests/drivers/uart_basic_api
Tested with previous reproducable issue:
west build -b up_squared tests/arch/x86/info/

Fixes #25260 (partial)

Signed-off-by: Jennifer Williams <jennifer.m.williams@intel.com>

pabigot reviewed (commented):
~removed partial review, can't delete thank you github~

pabigot reviewed (dismissed):
Trying again:  This doesn't fix the problem, which can be reproduced with:

    west build -b up_squared tests/arch/x86/info/

which produces:
```
/mnt/nordic/zp/zephyr/drivers/serial/uart_ns16550.c: In function 'uart_ns16550_configure':
/mnt/nordic/zp/zephyr/drivers/serial/uart_ns16550.c:333:25: error: assignment of member 'port' in read-only object
  333 |   dev_cfg->devconf.port = pcie_get_mbar(dev_cfg->pcie_bdf, 0);
      |    
```

The basic issue is that each driver's `struct uart_ns16550_device_config` is defined as `const` meaning it's placed in flash.  The pointer to it is marked `const void *` in `struct device`, but the original cast to the driver-specific type removed the `const` qualifier.

The code that now produces an error is trying to change a field in the config structure.  Though it appeared to work (because the compiler was told that the pointed-to object could be changed) it would never work unless `const` data on up_squared is actually not const.

The required fix is to move the `port` field from `struct uart_ns16550_device_config` to `struct uart_ns16550_dev_data_t`.

jenmwms commented:
@pabigot Thanks for your comment! It helped me. Please take another look, I believe I implemented (something like) what you suggested.

pabigot reviewed (commented):
This removes the attempt to write to const memory, but either it doesn't work, or the PCIE cache override of `port` is unnecessary.

I think you should go back to #21966 and look at what changed there.  The port used to be accessed through `DEV_DATA(dev)->port`; although what you have in this PR restores that field back and sets it, all the accessor macros like `IER(dev)` still use `DEV_CFG(dev)->devconf.port` which is initialized from devicetree and can't be changed.  If you fix that (i.e. switch all the uses back to `DEV_DATA`) then you'll need to set `DEV_DATA(dev)->port` from `DEV_CFG(dev)->devconf.port` in the init function.

Is this being tested on upsquared hardware?  I don't have that board, so can't confirm this works.

jenmwms commented:
> If you fix that (i.e. switch all the uses back to `DEV_DATA`) then you'll need to set `DEV_DATA(dev)->port` from `DEV_CFG(dev)->devconf.port` in the init function.

@pabigot  Thank you, that helps very much. Your suggestion changes many lines in the code, but it makes sense. I will give that a try and push the changes.

jenmwms commented:
> Is this being tested on upsquared hardware? I don't have that board, so can't confirm this works.

Ok, this should be better. However, despite passing on qemu_x86 and qemu_x86_64, this is failing for up_squared physical HW. I tested and this is the result:
`
error: no suitable video mode found.
`
and nothing else. I'll keep working on it.

andrewboie commented:
In the fullness of time, for the specific case of driver information like MMIO addresses we might need some additional infrastructure to optimize where this information is maintained, rather than having to choose whether to unconditionally put the info in dev data (RAM) or config struct (ROM).

In addition to these hot-plug situations where in some configurations MMIO is only known at runtime, there is virtual memory mappings to consider a la Linux `ioremap()`.

Some thoughts here: https://github.com/zephyrproject-rtos/zephyr/issues/25389

dcpleung commented on drivers/serial/uart_ns16550.c:
```diff
@@ -467,6 +470,8 @@ static int uart_ns16550_init(struct device *dev)
 	DEV_CFG(dev)->devconf.irq_config_func(dev);
 #endif
 
+	DEV_DATA(dev)->port = DEV_CFG(dev)->devconf.port;
```
When PCIE is enabled, `port` is already set inside `uart_ns16550_configure()`. Setting this here will override it.

I think it is probably better to move the the whole `pcie_probe()` part here before calling `uart_ns16550_configure()`. The port only needs to be probed once, and there is no need to probe PCIE every single time UART is being configured.

dcpleung replied:
This is probably why it didn't work on the UP2 board.

jenmwms replied:
@dcpleung Thank you! That's a great point. A note - if I remove this line 473, the applications which do not use PCIE do not get configured. So, I can specify #ifndef PCIE for this line, and I'll move the whole pcie_probe() here before calling _configure(). I'll push an update shortly!

jenmwms commented:
Ok, here is an update addressing the issue @dcpleung pointed out. This is passsing `uart_basic_api`  tests on `qemu_x86`, `qemu_x86_64`, and `up_squared`. 

Also, this passes `tests/arch/x86/info/` on `up_squared` physical HW @pabigot:

```
*** Booting Zephyr OS build zephyr-v2.2.0-2851-ge50347989c14  ***


info: the Zephyr x86 platform information tool

MULTIBOOT: boot info structure available.
        Flags = 0x00000265
        Basic memory map: lower = 252K, upper = 261120K.
        Extended memory map was at 0x0001008c (672 bytes).
        Framebuffer data not present.

MEMORY MAP: source is Multiboot extended map, max 64 entries
        0000000000000000 -> 000000000003efff RAM (252K)
        000000000003f000 -> 000000000003ffff UNKNOWN (4K)
        0000000000040000 -> 000000000009dfff RAM (376K)
        000000000009e000 -> 00000000000fffff UNKNOWN (392K)
        0000000000100000 -> 000000000fffffff RAM (261120K)
        0000000010000000 -> 0000000012150fff UNKNOWN (34116K)
        0000000012151000 -> 0000000077b0cfff RAM (1664752K)
        0000000077b0d000 -> 0000000079c2afff UNKNOWN (33912K)
        0000000079c2b000 -> 0000000079c43fff ACPI (100K)
        0000000079c44000 -> 0000000079ca3fff NVS (384K)
        0000000079ca4000 -> 000000007a040fff UNKNOWN (3700K)
        000000007a041000 -> 000000007a08cfff UNKNOWN (304K)
        000000007a08d000 -> 000000007a3fafff RAM (3512K)
        000000007a3fb000 -> 000000007a3fbfff NVS (4K)
        000000007a3fc000 -> 000000007a425fff UNKNOWN (168K)
        000000007a426000 -> 000000007a964fff RAM (5372K)
        000000007a965000 -> 000000007a966fff UNKNOWN (8K)
        000000007a967000 -> 000000007affffff RAM (6756K)
        000000007b000000 -> 000000007fffffff UNKNOWN (81920K)
        00000000d0000000 -> 00000000d0ffffff UNKNOWN (16384K)
        00000000e0000000 -> 00000000efffffff UNKNOWN (262144K)
        00000000fe042000 -> 00000000fe044fff UNKNOWN (12K)
        00000000fe900000 -> 00000000fe902fff UNKNOWN (12K)
        00000000fec00000 -> 00000000fec00fff UNKNOWN (4K)
        00000000fed01000 -> 00000000fed01fff UNKNOWN (4K)
        00000000fee00000 -> 00000000fee00fff UNKNOWN (4K)
        00000000ff000000 -> 00000000ffffffff UNKNOWN (16384K)
        0000000100000000 -> 000000017fffffff RAM (2097152K)

        Kernel exclusions:
        0x00100000 -> 0x00103c8b (rom)
        0x00103ca0 -> 0x00105cdc (ram)

ACPI: 4 CPUs found
        CPU #0: APIC ID 0x00
        CPU #1: APIC ID 0x02
        CPU #2: APIC ID 0x04
        CPU #3: APIC ID 0x06

TIMER: new local APIC, configured frequency = 19200000Hz
        Using CMOS RTC as reference clock:
        start = 244646539, end = 263847731, 19201192 cycles
        start = 283049027, end = 302249951, 19200924 cycles
        start = 321452411, end = 340652010, 19199599 cycles
        start = 359854380, end = 379054663, 19200283 cycles
        start = 398255703, end = 417457266, 19201563 cycles
        start = 436658409, end = 455860191, 19201782 cycles
        start = 475060556, end = 494262629, 19202073 cycles
        start = 513462743, end = 532664169, 19201426 cycles
        start = 551865534, end = 571067025, 19201491 cycles
        start = 590268810, end = 609469267, 19200457 cycles
        average = 19201079Hz

info: complete

```

jenmwms commented:
Hmm. I have a merge conflict detected by shippable. I'll rebase again and restart Shippable.

dcpleung commented on drivers/serial/uart_ns16550.c:
```diff
@@ -461,12 +453,33 @@ static int uart_ns16550_config_get(struct device *dev, struct uart_config *cfg)
 #endif
 
+#ifndef UART_NS16550_PCIE_ENABLED
+	DEV_DATA(dev)->port = DEV_CFG(dev)->devconf.port;
```
This is not right... If PCIE is not enabled, `DEV_DATA(dev)->port` is not set when calling `uart_ns16550_configure()` above. So none of the configuration is being done. And that would be accessing I/O port @ 0x0000, or memory @ 0x00000000 + offset.

jenmwms replied:
@dcpleung Thanks for looking at it. That is interesting. Perhaps that is because of the recent change implemented as @pabigot suggested as a resolution to the handling of port setting:
> If you fix that (i.e. switch all the uses back to DEV_DATA) then you'll need to set DEV_DATA(dev)->port from DEV_CFG(dev)->devconf.port in the init function.

jenmwms replied:
I'll need to look at this more

jenmwms commented:
CI checks are passing, but there is a conflicting file `drivers/serial/uart_ns16550.c`. I'll rebase and push again.

jenmwms committed:
drivers: uart: fix use of port / config info from const struct

This driver was assigning to a field that was stored in a const
structure. The issue was (mostly) resolved in af43e14 by moving
the field to a RAM structure and in #25462 by preserving the
original const qualifier. Moving the field was not implemented
for PCIE=n and introduced an inline function for the helper
macros that have otherwise been consistent.

This commit further addresses the issue by moving the port field for
PCIE=y and PCIE=n. It lets the get_port() workaround be refactored
back to DEV_DATA() for consistency in the helper macros. The PCIE
probe is moved to the init() function to avoid probing PCIE every
time UART is being configured.

Tested with qemu_x86 and qemu_x86_64 tests/drivers/uart_basic_api
Tested with previous reproducable issue:
west build -b up_squared tests/arch/x86/info/

Fixes #25260 (partial)

Signed-off-by: Jennifer Williams <jennifer.m.williams@intel.com>

jenmwms commented:
Ok, here is an update addressing the merge conflict. This is passing uart_basic_api tests on qemu_x86, qemu_x86_64, and up_squared.
Also, this passes tests/arch/x86/info/ on up_squared physical HW.

@dcpleung I recognize this updated patch keeps the whole pcie_probe() part in its existing spot in `_configure()`. It also has the `#ifndef` PCIE around the `DEV_DATA(dev)->devconf.port` assignment in `_init()`. I had issues with Up2 the other way, but I will keep looking into it - it should not stop release, but could be an enhancement yeah?

jenmwms commented:
fixed gitlint issue (1 single whitepsace...oops)

dcpleung commented:
Actually, overwrite the files with the changes before the rebase. Those are good changes. The only issue left is to assign the port address before calling `configure()` for non-PCIe cases.

jenmwms commented:
> Actually, overwrite the files with the changes before the rebase. Those are good changes. The only issue left is to assign the port address before calling `configure()` for non-PCIe cases.

Thanks for letting me know! Cool. I will work on that - something is different after making those changes the same way as before, because it does not work on up2 right away. More soon...

jenmwms commented:
ok, here is an update! This does not work on `up_squared` yet (suggesting it is related to my moving the `pcie_probe()` chunk from `_configure()` to here in `_init()`).  qemu_x86 and qemu_x86_64 work for `tests/drivers/uart/uart_basic_api`. More soon...

jenmwms commented:
In light of recent activity on #25260 - the issue is resolved by af43e14  and #25462 -  is this PR considered a bug fix or an enhancement?

jenmwms commented:
I'll rebase and push so we can see the up-to-date difference.

jenmwms commented:
> ok, here is an update! This does not work on `up_squared` yet (suggesting it is related to my moving the `pcie_probe()` chunk from `_configure()` to here in `_init()`). qemu_x86 and qemu_x86_64 work for `tests/drivers/uart/uart_basic_api`. More soon...

still working on with this after rebase.

nashif commented:
> In light of recent activity on #25260 - the issue is resolved by [af43e14](https://github.com/zephyrproject-rtos/zephyr/commit/af43e14bd0dbcdb0595b1e6620f85fd3513a9eaa) and #25462 - is this PR considered a bug fix or an enhancement?

afaik, this is still a bug, and the PR title does not reflect the changes in this PR, not sure why the original bug was closed, this is confusing. @pabigot ?

pabigot commented:
> > In light of recent activity on #25260 - the issue is resolved by [af43e14](https://github.com/zephyrproject-rtos/zephyr/commit/af43e14bd0dbcdb0595b1e6620f85fd3513a9eaa) and #25462 - is this PR considered a bug fix or an enhancement?
> 
> afaik, this is still a bug, and the PR title does not reflect the changes in this PR, not sure why the original bug was closed, this is confusing. @pabigot ?

@nashif  #25260 was opened because this driver was assigning to a field that was stored in a  `const` structure, which would never have worked.  I believe that has been fixed by moving the field to a RAM structure (af43e14bd0dbcdb0595b1e6620f85fd3513a9eaa) and correcting the helper macros to preserve the original `const` qualifier when casting a void pointer to the immutable configuration data (#25462).

The fix for the original problem in this PR was combined with other refactoring that is still present.

jenmwms commented:
Note; I meant to mention that I need to update the title and commit message based on the feedback if this is still addressing the bug or is an enhancement.

jenmwms commented:
Thanks for keeping this in consideration. I updated the PR and commit message to help reflect the clarifications. There were a few approaches underway for this issue without knowing, it happens. The faster fixes have been merged, but I believe this PR offers an improved/unified approach since we have been able to work on this together for some time now to review and test (WIP). The fixes were originally part of this PR that has developed into richer contribution that help better address the issue.

dcpleung commented on drivers/serial/uart_ns16550.c:
```diff
@@ -484,8 +468,33 @@ static int uart_ns16550_config_get(struct device *dev, struct uart_config *cfg)
+
+		dev_data->devconf.port = pcie_get_mbar(dev_cfg->pcie_bdf, 0);
+		pcie_set_cmd(dev_cfg->pcie_bdf, PCIE_CONF_CMDSTAT_MEM, true);
+	}
+#endif
+
+/*
+ * Set port to DEV_DATA, but not overwrite for PCIE enabled
+ * set by pcie_get_mbar() via _configure() above
+ */
+#ifndef UART_NS16550_PCIE_ENABLED
+	DEV_DATA(dev)->devconf.port = DEV_CFG(dev)->devconf.port;
+#endif
```
This might be better so it can accommodate platforms with mixed PCIE/non-PCIE UARTs:
```
        } else
#endif
        {
                DEV_DATA(dev)->devconf.port = DEV_CFG(dev)->devconf.port;
        {
```

jenmwms replied:
Much better, thanks!

jenmwms commented:
This does not work on up_squared yet. qemu_x86 and qemu_x86_64 work for tests/drivers/uart/uart_basic_api. More soon...ideas?

jenmwms commented:
The issue has been resolved by #25462 and changes to source code in the meanwhile have rendered this PR enhancements out of date. Let's close it and fresh look at enhancements can be suggested.
